storage/src/key_value/agent.rs

```diff
@@ -0,0 +1,603 @@
+        }
+        Ok(())
+    }
+
+    fn get_digest_keys(&mut self, column: KeyValueColumn) -> Result<Vec<Digest>> {
+        let keys = self.inner.get_all_keys(column)?;
+        let mut out = Vec::with_capacity(keys.len());
+        for key in keys {
+            out.push(key[..].into());
+        }
+        Ok(out)
+    }
+
+    fn wrap<T, F: FnOnce(&mut Self) -> Result<T>>(&mut self, func: F) -> Result<T> {
+        self.inner.begin()?;
+        let out = func(self);
+        if out.is_err() {
+            self.inner.abort()?;
+        } else {
+            self.inner.commit()?;
+        }
+        out
+    }
+
+    fn handle_message(&mut self, message: Message) -> Box<dyn Any + Send + Sync> {
+        match message {
+            Message::InsertBlock(block) => Box::new(self.wrap(move |f| f.insert_block(block))),
+            Message::GetBlockHash(block_num) => Box::new(self.get_block_hash(block_num)),
+            Message::GetBlockHeader(hash) => Box::new(self.get_block_header(&hash)),
+            Message::GetBlockState(hash) => Box::new(self.get_block_state(&hash)),
+            Message::GetBlockStates(hashes) => Box::new(self.get_block_states(&hashes[..])),
+            Message::GetBlock(hash) => Box::new(self.get_block(&hash)),
+            Message::GetForkPath(hash, oldest_fork_threshold) => {
+                Box::new(self.get_fork_path(&hash, oldest_fork_threshold))
+            }
+            Message::CommitBlock(block_hash, ledger_digest) => {
+                Box::new(self.wrap(move |f| f.commit_block(&block_hash, &ledger_digest)))
+            }
+            Message::DecommitBlocks(hash) => Box::new(self.wrap(move |f| f.decommit_blocks(&hash))),
+            Message::Canon() => Box::new(self.canon()),
+            Message::LongestChildPath(hash) => Box::new(self.longest_child_path(&hash)),
+            Message::GetBlockLocatorHashes() => Box::new(self.get_block_locator_hashes()),
+            Message::FindSyncBlocks(hashes, block_count) => Box::new(self.find_sync_blocks(hashes, block_count)),
+            Message::GetTransactionLocation(transaction_id) => Box::new(self.get_transaction_location(&transaction_id)),
+            Message::GetRecordCommitments(limit) => Box::new(self.get_record_commitments(limit)),
+            Message::GetRecord(commitment) => Box::new(self.get_record(&commitment)),
+            Message::StoreRecords(records) => Box::new(self.wrap(move |f| f.store_records(records))),
+            Message::GetCommitments() => Box::new(self.get_digest_keys(KeyValueColumn::Commitment)),
+            Message::GetSerials() => Box::new(self.get_digest_keys(KeyValueColumn::SerialNumber)),
+            Message::GetMemos() => Box::new(self.get_digest_keys(KeyValueColumn::Memo)),
+            Message::GetLedgerDigests() => Box::new(self.get_digest_keys(KeyValueColumn::DigestIndex)),
+        }
+    }
+
+    pub(super) fn agent(mut self, mut receiver: mpsc::Receiver<MessageWrapper>) {
```

Comment:
Why not `send`? This name doesn't suggest what it's doing, especially since it's on an `Agent` object.

Reply:
it is an executor for the agent. this is a common pattern